TumbleBitSetup/PoupardStern.cs

```diff
@@ -0,0 +1,273 @@
+                zValues[i] =  SampleFromZnStar(pubKey, psBytes, i, BigK, keyLength);
+
+            for (;;)
+            {
+                // Generate r
+                GetR(keyLength, out BigInteger r);
+
+                for (int i = 0; i < BigK; i++)
+                    // Compute x_i
+                    xValues[i] = zValues[i].ModPow(r, Modulus);
+
+                // Compute w
+                GetW(pubKey, psBytes, xValues, k, keyLength, out BigInteger w);
+
+                // Compute y
+                y = r.Add(NsubPhi.Multiply(w));
+
+                // if y >= 2^{ |N| - 1 }
+                if (y.CompareTo(lowerLimit) >= 0)
+                    continue;
+                
+                // if y < 0
+                if (y.CompareTo(BigInteger.Zero) < 0)
+                    continue;
+
+                return new Tuple<BigInteger[], BigInteger>(xValues, y);
+            }
+            
+        }
+
+        /// <summary>
+        /// Verifying Algorithm specified in page 13 (3.3) of the setup
+        /// </summary>
+        /// <param name="pubKey">Public key used</param>
+        /// <param name="xValues">List of x_i values</param>
+        /// <param name="y">The value y as specified in the setup</param>
+        /// <param name="keyLength">The size of the RSA key in bits</param>
+        /// <param name="ps">public string specified in the setup</param>
+        /// <param name="k">Security parameter specified in the setup.</param>
+        /// <returns>true if the xValues verify, false otherwise</returns>
+        public static bool Verifying(RsaPubKey pubKey, BigInteger[] xValues, BigInteger y, int keyLength, byte[] psBytes, int k = 128)
+        {
+            k = Utils.GetByteLength(k) * 8;
+            BigInteger rPrime;
+            BigInteger lowerLimit = BigInteger.Two.Pow(keyLength - 1);
+            BigInteger upperLimit = BigInteger.Two.Pow(keyLength);
+
+            var Modulus = pubKey._pubKey.Modulus;
+            var Exponent = pubKey._pubKey.Exponent;
+
+            // Checking that:
+            // if y >= 2^{ |N| - 1 }
+            if (y.CompareTo(lowerLimit) >= 0)  
+                return false;
+            // if y < 0
+            if (y.CompareTo(BigInteger.Zero) < 0)
+                return false;
+            // if N <= 2^{KeySize-1}
+            if (Modulus.CompareTo(lowerLimit) <= 0)
+                return false;
+            // if N >= 2^{KeySize}
+            if (Modulus.CompareTo(upperLimit) >= 0)
+                return false;
+
+            // Computing K
+            GetK(k, out int BigK);
+
+            // Check if the number of x_values is not equal to K
+            if (xValues.Length != BigK)
+                return false;
+
+            // Get w
+            GetW(pubKey, psBytes, xValues, k, keyLength, out BigInteger w);
+
+            // Computing rPrime
+            rPrime = y.Subtract(Modulus.Multiply(w));
+
+            // Encrypting and verifying the signatures
```

Comment:
Can you change this comment to just "verifying the signatures" ?

The word "encrypting" there is misleading